### PR TITLE
libnet/ipam: Various clean-ups

### PR DESCRIPTION
**Pull request by akerouanton**
**- What I did**

_Consider reviewing per commit._

- Un-embed mutex from `addrSpace`
- Put `addrSpace` into a separate file
- Remove dead functions `DumpDatabase` and `NetworkRange`
- Split v4/v6 `addrSpace`
- Introduce two new structs, `PoolRequest` and `AllocatedPool`, to represent the in/out values of `RequestPool`
- Rewrote a few tests to make them more idiomatic
- Remove the global vars used to store the local and global address pools
- Move package `libnet/ipam` to `libnet/ipams/defaultipam` -- grouping all IPAM drivers under `libnet/ipams`

**- How to verify it**

Through tests.

### CODE REVIEW

**Review by robmry — COMMENTED** [review]
Much neater, nice!

**Review by robmry — APPROVED** [review]

**Comment by akerouanton** [comment]
Ah, just realized one commit message was wrong. That's now fixed.

File: libnetwork/ipamapi/contract.go [hunk 1]
```diff
@@ -63,6 +60,41 @@ type Ipam interface {
...
+	// PoolID represents the ID of the allocated pool. Its value is opaque and
+	// shouldn't be interpreted by anything but the IPAM driver that generated
+	// it.
+	PoolID string
+	// Pool is the allocated prefix.
+	Pool netip.Prefix
+	// Meta represents a list of extra IP addresses automatically reserved
+	// during the pool allocation. These are generally keyed by well-known
+	// strings defined in the netlabel package.
+	Meta map[string]string
```
Comment (corhere) [inline libnetwork/ipamapi/contract.go:95]: If the values are always IP addresses, why are they stringly-typed? Or is the doc comment misleadingly describing how the default IPAM driver happens to populate a generic bag of data with implementation-defined semantics?

File: libnetwork/ipams/defaultipam/allocator.go [hunk 2]
```diff
@@ -68,41 +124,35 @@ func (a *Allocator) GetDefaultAddressSpaces() (string, string, error) {
...
 	}
-
-	if addressSpace == "" {
-		return "", nil, nil, parseErr(ipamapi.ErrInvalidAddressSpace)
-	}
-	aSpace, err := a.getAddrSpace(addressSpace)
+	aSpace, err := a.getAddrSpace(req.AddressSpace, req.V6)
 	if err != nil {
-		return "", nil, nil, err
+		return ipamapi.AllocatedPool{}, types.InternalErrorf("invalid pool request: invalid address space")
```
Comment (corhere) [inline libnetwork/ipams/defaultipam/allocator.go:135]: I'm not super thrilled that the well-known ipamapi error values aren't being used, but they were not being wrapped properly for use with `errors.Is` so it's not a regression. And I have a feeling all this code is going to be replaced in the follow-up to this PR anyway.

File: libnetwork/ipams/defaultipam/allocator_test.go [hunk 3]
```diff
@@ -360,16 +357,16 @@ func TestPoolAllocationReuse(t *testing.T) {
 	// Verify that we don't see any repeat networks even though
 	// we have freed them.
 	seen := map[string]bool{}
-	for i := 0; i < nPools; i++ {
-		pool, nw, _, err := a.RequestPool(localAddressSpace, "", "", nil, false)
+	for i := 0; i < len(allocs); i++ {
```
Comment (corhere) [inline libnetwork/ipams/defaultipam/allocator_test.go:360]: What's this, C?

```suggestion
	for range allocs {
```

File: libnetwork/network.go [hunk 4]
```diff
@@ -1508,21 +1509,27 @@ func (n *Network) ipamAllocate() error {
 	return err
 }
 
-func (n *Network) requestPoolHelper(ipam ipamapi.Ipam, addressSpace, requestedPool, requestedSubPool string, options map[string]string, v6 bool) (poolID string, pool *net.IPNet, meta map[string]string, err error) {
+func (n *Network) requestPoolHelper(ipam ipamapi.Ipam, addressSpace, requestedPool, requestedSubPool string, options map[string]string, v6 bool) (string, *net.IPNet, map[string]string, error) {
```
Comment (corhere) [inline libnetwork/network.go:1512]: Why'd you remove the names of the return values? Those are good documentation!

File: libnetwork/ipams/windowsipam/windowsipam_other.go [hunk 5]
```diff
@@ -0,0 +1,10 @@
+//go:build !windows
+
+package windowsipam
+
+import "github.com/docker/docker/libnetwork/ipamapi"
+
+// Register is a no-op -- windowsipam is only supported on Windows.
+func Register(_ ipamapi.Registerer) error {
```
Comment (corhere) [inline libnetwork/ipams/windowsipam/windowsipam_other.go:8]:
```suggestion
func Register(ipamapi.Registerer) error {
```

File: libnetwork/cnmallocator/drivers_ipam.go [hunk 6]
```diff
@@ -33,20 +32,13 @@ func initIPAMDrivers(r ipamapi.Registerer, netConfig *networkallocator.Config) e
...
 	}
 
-	for _, fn := range [](func(ipamapi.Registerer) error){
-		builtinIpam.Register,
-		nullIpam.Register,
-	} {
-		if err := fn(r); err != nil {
-			return err
-		}
+	if err := ipams.Register(r, nil, []*ipamutils.NetworkToSplit(nil), addressPool); err != nil {
```
Comment (corhere) [inline libnetwork/cnmallocator/drivers_ipam.go:40]: What's with the typed-nil slice value?
```suggestion
	if err := ipams.Register(r, nil, nil, addressPool); err != nil {
```

File: libnetwork/controller.go [hunk 7]
```diff
@@ -132,7 +134,7 @@ func New(cfgOptions ...config.Option) (*Controller, error) {
 		return nil, err
 	}
 
-	if err := initIPAMDrivers(&c.ipamRegistry, c.cfg.PluginGetter, c.cfg.DefaultAddressPool); err != nil {
+	if err := ipams.Register(&c.ipamRegistry, c.cfg.PluginGetter, c.cfg.DefaultAddressPool, []*ipamutils.NetworkToSplit(nil)); err != nil {
```
Comment (corhere) [inline libnetwork/controller.go:137]:
```suggestion
	if err := ipams.Register(&c.ipamRegistry, c.cfg.PluginGetter, c.cfg.DefaultAddressPool, nil); err != nil {
```

File: libnetwork/drvregistry/ipams_test.go [hunk 8]
```diff
@@ -6,19 +6,16 @@ import (
...
 	is "gotest.tools/v3/assert/cmp"
 )
 
 func getNewIPAMs(t *testing.T) *IPAMs {
 	r := &IPAMs{}
 
-	assert.Assert(t, builtinIpam.Register(r))
-	assert.Assert(t, remoteIpam.Register(r, nil))
-	assert.Assert(t, nullIpam.Register(r))
+	assert.Assert(t, ipams.Register(r, nil, []*ipamutils.NetworkToSplit(nil), []*ipamutils.NetworkToSplit(nil)))
```
Comment (corhere) [inline libnetwork/drvregistry/ipams_test.go:18]:
```suggestion
	assert.Assert(t, ipams.Register(r, nil, nil, nil))
```

File: libnetwork/ipamutils/utils.go [hunk 9]
```diff
@@ -48,22 +48,6 @@ func init() {
...
-	if err != nil {
-		return err
-	}
-	predefinedGlobalScopeDefaultNetworks = defaultNetworks
-	return nil
-}
-
 // GetGlobalScopeDefaultNetworks returns a copy of the global-sopce network list.
 func GetGlobalScopeDefaultNetworks() []*net.IPNet {
 	mutex.Lock()
```
Comment (corhere) [inline libnetwork/ipamutils/utils.go:53]: There's no more mutable global state, which means there's no more need for this mutex.

**Review by corhere — APPROVED** [review]
What an improvement! It's amazing how much we were held back by cnmallocator living in the Swarmkit repo.

The meat of the changes look good. All I really have are aesthetic quibbles; nothing that couldn't be ignored, or dealt with in a follow-up PR.

File: libnetwork/ipamapi/contract.go (hunk 1 above)
Comment (akerouanton) [reply libnetwork/ipamapi/contract.go:95]: Yeah, I'll update this comment -- it's misleading.

I'm also wondering how useful it is. No builtin driver set any `Meta`, only remote drivers might do that. And it's used only in two places:

https://github.com/moby/moby/blob/16b2c223ce1c2381b6be67aaa218f3b02746e01c/libnetwork/cnmallocator/networkallocator.go#L905-L910

https://github.com/moby/moby/blob/16b2c223ce1c2381b6be67aaa218f3b02746e01c/libnetwork/network.go#L1619-L1623

So, I think it'd even be safe to replace it with an optional `Gateway`.

**Review by robmry — APPROVED** [review]

**Comment by akerouanton** [comment]
Okay, let's merge this one and rebase the #47768.

**Pull request merged**